From 7346775e5d29cf3d39e4531b9640e73e285089c4 Mon Sep 17 00:00:00 2001
From: Steve Kondik <shade@chemlab.org>
Date: Sun, 7 Oct 2012 13:00:42 -0700
Subject: [PATCH 03/18] latinime: Add "more" keys to ALL the keys

 * I'm in ur productivity, increasin' it.

Change-Id: I688bed9c3db7ef8d2c9dd5e4af72a65eb7f37966

Unbreak QWERTZ keyboards

Adding "more" keys to ALL the keys broke QWERTZ keyboards.
Fix it by adding seperate XMLs for QWERTZ.

Change-Id: Iccf10139cf02cab2a773c7085aadac8538359a16

LatinIME: Add "more" keys to ALL the keys in Greek layouts

Change-Id: I120e7b94abd71baf24c97afbd20cf2c6c76df6ef

LatinIME: Apply "Enable more keys for top row on tablets" (f6620308ba)
and "Add "more" keys to ALL the keys" (c561f31af9) to slavik layouts.

Change-Id: Ic6fea3faecc5783d99f5257a4501e37b79f40fdf

Rearrange morekeys

  Some of the symbols such as *,!,-,+ were moved in kitkat,
  this moves the moreKeys to be inline with the moves.

Change-Id: I5aff699b268dd32a5b1e4539163e6ec153021f0a

Rearrange Slavic & Greek morekeys

* Following the previous change

Change-Id: I5b5a9bff8dd97c0bf31bc5e28245a9c14662d583
---
 java/res/xml-sw600dp/rows_qwertz.xml    |  2 +-
 java/res/xml/rowkeys_east_slavic2.xml   | 40 +++++++++++++++++-----
 java/res/xml/rowkeys_east_slavic3.xml   | 36 +++++++++++++++-----
 java/res/xml/rowkeys_greek2.xml         | 36 +++++++++++++++-----
 java/res/xml/rowkeys_greek3.xml         | 28 ++++++++++++----
 java/res/xml/rowkeys_qwerty2_left5.xml  | 12 ++++++-
 java/res/xml/rowkeys_qwerty2_right4.xml |  8 +++++
 java/res/xml/rowkeys_qwerty3_left4.xml  |  8 +++++
 java/res/xml/rowkeys_qwerty3_right3.xml | 10 ++++--
 java/res/xml/rowkeys_qwertz2.xml        | 30 +++++++++++++++++
 java/res/xml/rowkeys_qwertz2_left5.xml  | 44 +++++++++++++++++++++++++
 java/res/xml/rowkeys_qwertz2_right4.xml | 44 +++++++++++++++++++++++++
 java/res/xml/rowkeys_qwertz3.xml        | 20 +++++++++--
 java/res/xml/rows_qwertz.xml            |  2 +-
 14 files changed, 278 insertions(+), 42 deletions(-)
 create mode 100644 java/res/xml/rowkeys_qwertz2.xml
 create mode 100644 java/res/xml/rowkeys_qwertz2_left5.xml
 create mode 100644 java/res/xml/rowkeys_qwertz2_right4.xml

diff --git a/java/res/xml-sw600dp/rows_qwertz.xml b/java/res/xml-sw600dp/rows_qwertz.xml
index 3b59dec54..747dafdce 100644
--- a/java/res/xml-sw600dp/rows_qwertz.xml
+++ b/java/res/xml-sw600dp/rows_qwertz.xml
@@ -36,7 +36,7 @@
         latin:keyWidth="9.0%p"
     >
         <include
-            latin:keyboardLayout="@xml/rowkeys_qwerty2"
+            latin:keyboardLayout="@xml/rowkeys_qwertz2"
             latin:keyXPos="4.5%p" />
         <Key
             latin:keyStyle="enterKeyStyle"
diff --git a/java/res/xml/rowkeys_east_slavic2.xml b/java/res/xml/rowkeys_east_slavic2.xml
index 21463fb7d..4047ab7ba 100644
--- a/java/res/xml/rowkeys_east_slavic2.xml
+++ b/java/res/xml/rowkeys_east_slavic2.xml
@@ -23,37 +23,59 @@
 >
     <!-- U+0444: "ф" CYRILLIC SMALL LETTER EF -->
     <Key
-        latin:keySpec="&#x0444;" />
+        latin:keySpec="&#x0444;"
+        latin:keyHintLabel="\@"
+        latin:additionalMoreKeys="\\@" />
     <Key
         latin:keySpec="!text/keyspec_east_slavic_row2_2"
+        latin:keyHintLabel="#"
+        latin:additionalMoreKeys="#"
         latin:moreKeys="!text/morekeys_east_slavic_row2_2" />
     <!-- U+0432: "в" CYRILLIC SMALL LETTER VE -->
     <Key
-        latin:keySpec="&#x0432;" />
+        latin:keySpec="&#x0432;"
+        latin:keyHintLabel="$"
+        latin:additionalMoreKeys="$" />
     <!-- U+0430: "а" CYRILLIC SMALL LETTER A -->
     <Key
         latin:keySpec="&#x0430;"
-        latin:moreKeys="!text/morekeys_cyrillic_a" />
+        latin:moreKeys="!text/morekeys_cyrillic_a"
+        latin:keyHintLabel="%"
+        latin:additionalMoreKeys="%" />
     <!-- U+043F: "п" CYRILLIC SMALL LETTER PE -->
     <Key
-        latin:keySpec="&#x043F;" />
+        latin:keySpec="&#x043F;"
+        latin:keyHintLabel="&amp;"
+        latin:additionalMoreKeys="&amp;" />
     <!-- U+0440: "р" CYRILLIC SMALL LETTER ER -->
     <Key
-        latin:keySpec="&#x0440;" />
+        latin:keySpec="&#x0440;"
+        latin:keyHintLabel="_"
+        latin:additionalMoreKeys="_" />
     <!-- U+043E: "о" CYRILLIC SMALL LETTER O -->
     <Key
         latin:keySpec="&#x043E;"
+        latin:keyHintLabel="/"
+        latin:additionalMoreKeys="/"
         latin:moreKeys="!text/morekeys_cyrillic_o" />
     <!-- U+043B: "л" CYRILLIC SMALL LETTER EL -->
     <Key
-        latin:keySpec="&#x043B;" />
+        latin:keySpec="&#x043B;"
+        latin:keyHintLabel="-"
+        latin:additionalMoreKeys="-" />
     <!-- U+0434: "д" CYRILLIC SMALL LETTER DE -->
     <Key
-        latin:keySpec="&#x0434;" />
+        latin:keySpec="&#x0434;"
+        latin:keyHintLabel="+"
+        latin:additionalMoreKeys="+" />
     <!-- U+0436: "ж" CYRILLIC SMALL LETTER ZHE -->
     <Key
-        latin:keySpec="&#x0436;" />
+        latin:keySpec="&#x0436;"
+        latin:keyHintLabel="("
+        latin:additionalMoreKeys="(" />
     <Key
         latin:keySpec="!text/keyspec_east_slavic_row2_11"
-        latin:moreKeys="!text/morekeys_east_slavic_row2_11" />
+        latin:moreKeys="!text/morekeys_east_slavic_row2_11"
+        latin:keyHintLabel=")"
+        latin:additionalMoreKeys=")" />
 </merge>
diff --git a/java/res/xml/rowkeys_east_slavic3.xml b/java/res/xml/rowkeys_east_slavic3.xml
index 54802e878..e1b9b60e2 100644
--- a/java/res/xml/rowkeys_east_slavic3.xml
+++ b/java/res/xml/rowkeys_east_slavic3.xml
@@ -23,29 +23,47 @@
 >
     <!-- U+044F: "я" CYRILLIC SMALL LETTER YA -->
     <Key
-        latin:keySpec="&#x044F;" />
+        latin:keySpec="&#x044F;"
+        latin:keyHintLabel="*"
+        latin:additionalMoreKeys="*" />
     <!-- U+0447: "ч" CYRILLIC SMALL LETTER CHE -->
     <Key
-        latin:keySpec="&#x0447;" />
+        latin:keySpec="&#x0447;"
+        latin:keyHintLabel="&quot;"
+        latin:additionalMoreKeys="&quot;" />
     <!-- U+0441: "с" CYRILLIC SMALL LETTER ES -->
     <Key
-        latin:keySpec="&#x0441;" />
+        latin:keySpec="&#x0441;"
+        latin:keyHintLabel="&apos;"
+        latin:additionalMoreKeys="&apos;" />
     <!-- U+043C: "м" CYRILLIC SMALL LETTER EM -->
     <Key
-        latin:keySpec="&#x043C;" />
+        latin:keySpec="&#x043C;"
+        latin:keyHintLabel=":"
+        latin:additionalMoreKeys=":" />
     <Key
-        latin:keySpec="!text/keyspec_east_slavic_row3_5" />
+        latin:keySpec="!text/keyspec_east_slavic_row3_5"
+        latin:keyHintLabel=";"
+        latin:additionalMoreKeys=";" />
     <!-- U+0442: "т" CYRILLIC SMALL LETTER TE -->
     <Key
-        latin:keySpec="&#x0442;" />
+        latin:keySpec="&#x0442;"
+        latin:keyHintLabel="!"
+        latin:additionalMoreKeys="!" />
     <!-- U+044C: "ь" CYRILLIC SMALL LETTER SOFT SIGN -->
     <Key
         latin:keySpec="&#x044C;"
-        latin:moreKeys="!text/morekeys_cyrillic_soft_sign" />
+        latin:moreKeys="!text/morekeys_cyrillic_soft_sign"
+        latin:keyHintLabel="\?"
+        latin:additionalMoreKeys="\\?" />
     <!-- U+0431: "б" CYRILLIC SMALL LETTER BE -->
     <Key
-        latin:keySpec="&#x0431;" />
+        latin:keySpec="&#x0431;"
+        latin:keyHintLabel="&lt;"
+        latin:additionalMoreKeys="&lt;" />
     <!-- U+044E: "ю" CYRILLIC SMALL LETTER YU -->
     <Key
-        latin:keySpec="&#x044E;" />
+        latin:keySpec="&#x044E;"
+        latin:keyHintLabel="&gt;"
+        latin:additionalMoreKeys="&gt;" />
 </merge>
diff --git a/java/res/xml/rowkeys_greek2.xml b/java/res/xml/rowkeys_greek2.xml
index d8769caa9..ba2e9c697 100644
--- a/java/res/xml/rowkeys_greek2.xml
+++ b/java/res/xml/rowkeys_greek2.xml
@@ -25,31 +25,49 @@
          U+03AC: "ά" GREEK SMALL LETTER ALPHA WITH TONOS -->
     <Key
         latin:keySpec="&#x03B1;"
-        latin:moreKeys="&#x03AC;" />
+        latin:moreKeys="&#x03AC;"
+        latin:keyHintLabel="\@"
+        latin:additionalMoreKeys="\\@" />
     <!-- U+03C3: "σ" GREEK SMALL LETTER SIGMA -->
     <Key
-        latin:keySpec="&#x03C3;" />
+        latin:keySpec="&#x03C3;"
+        latin:keyHintLabel="#"
+        latin:additionalMoreKeys="#" />
     <!-- U+03B4: "δ" GREEK SMALL LETTER DELTA -->
     <Key
-        latin:keySpec="&#x03B4;" />
+        latin:keySpec="&#x03B4;"
+        latin:keyHintLabel="€"
+        latin:additionalMoreKeys="€" />
     <!-- U+03C6: "φ" GREEK SMALL LETTER PHI -->
     <Key
-        latin:keySpec="&#x03C6;" />
+        latin:keySpec="&#x03C6;"
+        latin:keyHintLabel="%"
+        latin:additionalMoreKeys="%" />
     <!-- U+03B3: "γ" GREEK SMALL LETTER GAMMA -->
     <Key
-        latin:keySpec="&#x03B3;" />
+        latin:keySpec="&#x03B3;"
+        latin:keyHintLabel="&amp;"
+        latin:additionalMoreKeys="&amp;" />
     <!-- U+03B7: "η" GREEK SMALL LETTER ETA
          U+03AE: "ή" GREEK SMALL LETTER ETA WITH TONOS -->
     <Key
         latin:keySpec="&#x03B7;"
-        latin:moreKeys="&#x03AE;" />
+        latin:moreKeys="&#x03AE;"
+        latin:keyHintLabel="-"
+        latin:additionalMoreKeys="-" />
     <!-- U+03BE: "ξ" GREEK SMALL LETTER XI -->
     <Key
-        latin:keySpec="&#x03BE;" />
+        latin:keySpec="&#x03BE;"
+        latin:keyHintLabel="+"
+        latin:additionalMoreKeys="+" />
     <!-- U+03BA: "κ" GREEK SMALL LETTER KAPPA -->
     <Key
-        latin:keySpec="&#x03BA;" />
+        latin:keySpec="&#x03BA;"
+        latin:keyHintLabel="("
+        latin:additionalMoreKeys="(" />
     <!-- U+03BB: "λ" GREEK SMALL LETTER LAMDA -->
     <Key
-        latin:keySpec="&#x03BB;" />
+        latin:keySpec="&#x03BB;"
+        latin:keyHintLabel=")"
+        latin:additionalMoreKeys=")" />
 </merge>
diff --git a/java/res/xml/rowkeys_greek3.xml b/java/res/xml/rowkeys_greek3.xml
index 3f989bc9c..cda9106b9 100644
--- a/java/res/xml/rowkeys_greek3.xml
+++ b/java/res/xml/rowkeys_greek3.xml
@@ -23,25 +23,39 @@
 >
     <!-- U+03B6: "ζ" GREEK SMALL LETTER ZETA -->
     <Key
-        latin:keySpec="&#x03B6;" />
+        latin:keySpec="&#x03B6;"
+        latin:keyHintLabel="*"
+        latin:additionalMoreKeys="*" />
     <!-- U+03C7: "χ" GREEK SMALL LETTER CHI -->
     <Key
-        latin:keySpec="&#x03C7;" />
+        latin:keySpec="&#x03C7;"
+        latin:keyHintLabel="&quot;"
+        latin:additionalMoreKeys="&quot;" />
     <!-- U+03C8: "ψ" GREEK SMALL LETTER PSI -->
     <Key
-        latin:keySpec="&#x03C8;" />
+        latin:keySpec="&#x03C8;"
+        latin:keyHintLabel="&apos;"
+        latin:additionalMoreKeys="&apos;" />
     <!-- U+03C9: "ω" GREEK SMALL LETTER OMEGA
          U+03CE: "ώ" GREEK SMALL LETTER OMEGA WITH TONOS -->
     <Key
         latin:keySpec="&#x03C9;"
-        latin:moreKeys="&#x03CE;" />
+        latin:moreKeys="&#x03CE;"
+        latin:keyHintLabel=":"
+        latin:additionalMoreKeys=":" />
     <!-- U+03B2: "β" GREEK SMALL LETTER BETA -->
     <Key
-        latin:keySpec="&#x03B2;" />
+        latin:keySpec="&#x03B2;"
+        latin:keyHintLabel=";"
+        latin:additionalMoreKeys=";" />
     <!-- U+03BD: "ν" GREEK SMALL LETTER NU -->
     <Key
-        latin:keySpec="&#x03BD;" />
+        latin:keySpec="&#x03BD;"
+        latin:keyHintLabel="!"
+        latin:additionalMoreKeys="!" />
     <!-- U+03BC: "μ" GREEK SMALL LETTER MU -->
     <Key
-        latin:keySpec="&#x03BC;" />
+        latin:keySpec="&#x03BC;"
+        latin:keyHintLabel="\?"
+        latin:additionalMoreKeys="\\?" />
 </merge>
diff --git a/java/res/xml/rowkeys_qwerty2_left5.xml b/java/res/xml/rowkeys_qwerty2_left5.xml
index 1803bf203..3f8517c92 100644
--- a/java/res/xml/rowkeys_qwerty2_left5.xml
+++ b/java/res/xml/rowkeys_qwerty2_left5.xml
@@ -23,16 +23,26 @@
 >
     <Key
         latin:keySpec="a"
+        latin:keyHintLabel="\@"
+        latin:additionalMoreKeys="\\@"
         latin:moreKeys="!text/morekeys_a" />
     <Key
         latin:keySpec="s"
+        latin:keyHintLabel="#"
+        latin:additionalMoreKeys="#"
         latin:moreKeys="!text/morekeys_s" />
     <Key
         latin:keySpec="d"
+        latin:keyHintLabel="$"
+        latin:additionalMoreKeys="$"
         latin:moreKeys="!text/morekeys_d" />
     <Key
-        latin:keySpec="f" />
+        latin:keySpec="f"
+        latin:keyHintLabel="%"
+        latin:additionalMoreKeys="%" />
     <Key
         latin:keySpec="g"
+        latin:keyHintLabel="&amp;"
+        latin:additionalMoreKeys="&amp;"
         latin:moreKeys="!text/morekeys_g" />
 </merge>
diff --git a/java/res/xml/rowkeys_qwerty2_right4.xml b/java/res/xml/rowkeys_qwerty2_right4.xml
index 99936b7a2..c936280d7 100644
--- a/java/res/xml/rowkeys_qwerty2_right4.xml
+++ b/java/res/xml/rowkeys_qwerty2_right4.xml
@@ -23,14 +23,22 @@
 >
     <Key
         latin:keySpec="h"
+        latin:keyHintLabel="-"
+        latin:additionalMoreKeys="-"
         latin:moreKeys="!text/morekeys_h" />
     <Key
         latin:keySpec="j"
+        latin:keyHintLabel="+"
+        latin:additionalMoreKeys="+"
         latin:moreKeys="!text/morekeys_j" />
     <Key
         latin:keySpec="k"
+        latin:keyHintLabel="("
+        latin:additionalMoreKeys="("
         latin:moreKeys="!text/morekeys_k" />
     <Key
         latin:keySpec="l"
+        latin:keyHintLabel=")"
+        latin:additionalMoreKeys=")"
         latin:moreKeys="!text/morekeys_l" />
 </merge>
diff --git a/java/res/xml/rowkeys_qwerty3_left4.xml b/java/res/xml/rowkeys_qwerty3_left4.xml
index 6043c3bba..8dc9d41c5 100644
--- a/java/res/xml/rowkeys_qwerty3_left4.xml
+++ b/java/res/xml/rowkeys_qwerty3_left4.xml
@@ -23,14 +23,22 @@
 >
     <Key
         latin:keySpec="z"
+        latin:keyHintLabel="*"
+        latin:additionalMoreKeys="*"
         latin:moreKeys="!text/morekeys_z" />
     <Key
         latin:keySpec="!text/keyspec_x"
+        latin:keyHintLabel="&quot;"
+        latin:additionalMoreKeys="&quot;"
         latin:moreKeys="!text/morekeys_x" />
     <Key
         latin:keySpec="c"
+        latin:keyHintLabel="&apos;"
+        latin:additionalMoreKeys="&apos;"
         latin:moreKeys="!text/morekeys_c" />
     <Key
         latin:keySpec="v"
+        latin:keyHintLabel=":"
+        latin:additionalMoreKeys=":"
         latin:moreKeys="!text/morekeys_v" />
 </merge>
diff --git a/java/res/xml/rowkeys_qwerty3_right3.xml b/java/res/xml/rowkeys_qwerty3_right3.xml
index f69910344..464602907 100644
--- a/java/res/xml/rowkeys_qwerty3_right3.xml
+++ b/java/res/xml/rowkeys_qwerty3_right3.xml
@@ -22,10 +22,16 @@
     xmlns:latin="http://schemas.android.com/apk/res/com.android.inputmethod.latin"
 >
     <Key
-        latin:keySpec="b" />
+        latin:keySpec="b"
+        latin:keyHintLabel=";"
+        latin:additionalMoreKeys=";" />
     <Key
         latin:keySpec="n"
+        latin:keyHintLabel="!"
+        latin:additionalMoreKeys="!"
         latin:moreKeys="!text/morekeys_n" />
     <Key
-        latin:keySpec="m" />
+        latin:keySpec="m"
+        latin:keyHintLabel="\?"
+        latin:additionalMoreKeys="\\?" />
 </merge>
diff --git a/java/res/xml/rowkeys_qwertz2.xml b/java/res/xml/rowkeys_qwertz2.xml
new file mode 100644
index 000000000..10878a019
--- /dev/null
+++ b/java/res/xml/rowkeys_qwertz2.xml
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+/*
+**
+** Copyright 2012, The Android Open Source Project
+**
+** Licensed under the Apache License, Version 2.0 (the "License");
+** you may not use this file except in compliance with the License.
+** You may obtain a copy of the License at
+**
+**     http://www.apache.org/licenses/LICENSE-2.0
+**
+** Unless required by applicable law or agreed to in writing, software
+** distributed under the License is distributed on an "AS IS" BASIS,
+** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+** See the License for the specific language governing permissions and
+** limitations under the License.
+*/
+-->
+
+<merge
+    xmlns:latin="http://schemas.android.com/apk/res/com.android.inputmethod.latin"
+>
+    <!-- a,s,d,f,g -->
+    <include
+        latin:keyboardLayout="@xml/rowkeys_qwertz2_left5" />
+    <!-- h,j,k,l -->
+    <include
+        latin:keyboardLayout="@xml/rowkeys_qwertz2_right4" />
+</merge>
diff --git a/java/res/xml/rowkeys_qwertz2_left5.xml b/java/res/xml/rowkeys_qwertz2_left5.xml
new file mode 100644
index 000000000..f0ccaa508
--- /dev/null
+++ b/java/res/xml/rowkeys_qwertz2_left5.xml
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+/*
+**
+** Copyright 2014, The Android Open Source Project
+**
+** Licensed under the Apache License, Version 2.0 (the "License");
+** you may not use this file except in compliance with the License.
+** You may obtain a copy of the License at
+**
+**     http://www.apache.org/licenses/LICENSE-2.0
+**
+** Unless required by applicable law or agreed to in writing, software
+** distributed under the License is distributed on an "AS IS" BASIS,
+** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+** See the License for the specific language governing permissions and
+** limitations under the License.
+*/
+-->
+
+<merge
+    xmlns:latin="http://schemas.android.com/apk/res/com.android.inputmethod.latin"
+>
+    <Key
+        latin:keySpec="a"
+        latin:moreKeys="!text/morekeys_a" />
+    <Key
+        latin:keySpec="s"
+        latin:moreKeys="!text/morekeys_s" />
+    <Key
+        latin:keySpec="d"
+        latin:keyHintLabel="€"
+        latin:additionalMoreKeys="€"
+        latin:moreKeys="!text/morekeys_d" />
+    <Key
+        latin:keySpec="f"
+        latin:keyHintLabel="%"
+        latin:additionalMoreKeys="%" />
+    <Key
+        latin:keySpec="g"
+        latin:keyHintLabel="&amp;"
+        latin:additionalMoreKeys="&amp;"
+        latin:moreKeys="!text/morekeys_g" />
+</merge>
diff --git a/java/res/xml/rowkeys_qwertz2_right4.xml b/java/res/xml/rowkeys_qwertz2_right4.xml
new file mode 100644
index 000000000..c936280d7
--- /dev/null
+++ b/java/res/xml/rowkeys_qwertz2_right4.xml
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+/*
+**
+** Copyright 2014, The Android Open Source Project
+**
+** Licensed under the Apache License, Version 2.0 (the "License");
+** you may not use this file except in compliance with the License.
+** You may obtain a copy of the License at
+**
+**     http://www.apache.org/licenses/LICENSE-2.0
+**
+** Unless required by applicable law or agreed to in writing, software
+** distributed under the License is distributed on an "AS IS" BASIS,
+** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+** See the License for the specific language governing permissions and
+** limitations under the License.
+*/
+-->
+
+<merge
+    xmlns:latin="http://schemas.android.com/apk/res/com.android.inputmethod.latin"
+>
+    <Key
+        latin:keySpec="h"
+        latin:keyHintLabel="-"
+        latin:additionalMoreKeys="-"
+        latin:moreKeys="!text/morekeys_h" />
+    <Key
+        latin:keySpec="j"
+        latin:keyHintLabel="+"
+        latin:additionalMoreKeys="+"
+        latin:moreKeys="!text/morekeys_j" />
+    <Key
+        latin:keySpec="k"
+        latin:keyHintLabel="("
+        latin:additionalMoreKeys="("
+        latin:moreKeys="!text/morekeys_k" />
+    <Key
+        latin:keySpec="l"
+        latin:keyHintLabel=")"
+        latin:additionalMoreKeys=")"
+        latin:moreKeys="!text/morekeys_l" />
+</merge>
diff --git a/java/res/xml/rowkeys_qwertz3.xml b/java/res/xml/rowkeys_qwertz3.xml
index a66c34868..18c60b3a6 100644
--- a/java/res/xml/rowkeys_qwertz3.xml
+++ b/java/res/xml/rowkeys_qwertz3.xml
@@ -23,20 +23,34 @@
 >
     <Key
         latin:keySpec="y"
+        latin:keyHintLabel="*"
+        latin:additionalMoreKeys="*"
         latin:moreKeys="!text/morekeys_y" />
     <Key
-        latin:keySpec="x" />
+        latin:keySpec="x"
+        latin:keyHintLabel="&quot;"
+        latin:additionalMoreKeys="&quot;" />
     <Key
         latin:keySpec="c"
+        latin:keyHintLabel="&apos;"
+        latin:additionalMoreKeys="&apos;"
         latin:moreKeys="!text/morekeys_c" />
     <Key
         latin:keySpec="v"
+        latin:keyHintLabel=":"
+        latin:additionalMoreKeys=":"
         latin:moreKeys="!text/morekeys_v" />
     <Key
-        latin:keySpec="b" />
+        latin:keySpec="b"
+        latin:keyHintLabel=";"
+        latin:additionalMoreKeys=";" />
     <Key
         latin:keySpec="n"
+        latin:keyHintLabel="!"
+        latin:additionalMoreKeys="!"
         latin:moreKeys="!text/morekeys_n" />
     <Key
-        latin:keySpec="m" />
+        latin:keySpec="m"
+        latin:keyHintLabel="\?"
+        latin:additionalMoreKeys="\\?" />
 </merge>
diff --git a/java/res/xml/rows_qwertz.xml b/java/res/xml/rows_qwertz.xml
index 31a147c12..e4ba99fbe 100644
--- a/java/res/xml/rows_qwertz.xml
+++ b/java/res/xml/rows_qwertz.xml
@@ -33,7 +33,7 @@
         latin:keyWidth="10%p"
     >
         <include
-            latin:keyboardLayout="@xml/rowkeys_qwerty2"
+            latin:keyboardLayout="@xml/rowkeys_qwertz2"
             latin:keyXPos="5%p" />
     </Row>
     <Row
-- 
2.17.1

